### PR TITLE
Cluster managers ACL

### PR DESCRIPTION
**Pull request by khssnv**
New storage item for cluster managers ACL with extrinsics to manage it by governance (sudo).

### CODE REVIEW

File: pallets/ddc-staking/src/benchmarking.rs [hunk 1]
```diff
@@ -105,4 +105,22 @@ benchmarks! {
 	verify {
 		assert!(Ledger::<T>::contains_key(&new_controller));
 	}
+
+	allow_cluster_manager {
```
Comment (MRamanenkau) [inline pallets/ddc-staking/src/benchmarking.rs:109]: Do you have plans to extend the cluster manager logic? If so, having a separate pallet for that would be nice.

File: pallets/ddc-staking/src/tests.rs [hunk 2]
```diff
@@ -176,3 +178,32 @@ fn staking_should_work() {
 		assert_eq!(DdcStaking::edges(3), None);
 	});
 }
+
+#[test]
+fn cluster_managers_list_can_be_managed_by_governance_only() {
+	ExtBuilder::default().build_and_execute(|| {
+		// Governance can allow an account to become cluster manager.
+		assert_ok!(DdcStaking::allow_cluster_manager(RuntimeOrigin::root(), 1));
```
Comment (MRamanenkau) [inline pallets/ddc-staking/src/tests.rs:186]: Why `governance` if the origin is root?

Comment (khssnv) [reply pallets/ddc-staking/src/tests.rs:186]: As far as I understand governance (sudo, referenda) decisions are dispatched from the root origin.
https://github.com/search?q=repo%3Aparitytech%2Fpolkadot-sdk%20frame_system%3A%3ARawOrigin%3A%3ARoot&type=code

File: pallets/ddc-staking/src/benchmarking.rs (hunk 1 above)
Comment (khssnv) [reply pallets/ddc-staking/src/benchmarking.rs:109]: Planned features are about staking for cluster managers which will be similar to existing roles (edge, storage). Looks like at least staking and participation-related logic could live here.

File: pallets/ddc-staking/src/lib.rs [hunk 3]
```diff
@@ -861,6 +866,46 @@ pub mod pallet {
...
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let grantee = T::Lookup::lookup(grantee)?;
+			ClusterManagers::<T>::mutate(|grantees| {
+				if !grantees.contains(&grantee) {
+					grantees.push(grantee);
+				}
+			});
+
```
Comment (MRamanenkau) [inline pallets/ddc-staking/src/lib.rs:886]: Usually, it's a good practice to `deposit_event` before the return. At least to make it visible in the explorer.

File: pallets/ddc-staking/src/lib.rs [hunk 4]
```diff
@@ -861,6 +866,46 @@ pub mod pallet {
...
+		/// Add a new account to the list of cluster managers.
+		///
+		/// RuntimeOrigin must be Root to call this function.
+		#[pallet::weight(T::WeightInfo::allow_cluster_manager())]
+		pub fn allow_cluster_manager(
+			origin: OriginFor<T>,
+			grantee: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			ensure_root(origin)?;
```
Comment (Raid5594) [inline pallets/ddc-staking/src/lib.rs:878]: Do want to have the root defining cluster managers in mainnet as well?
Maybe there should be some governance process

File: pallets/ddc-staking/src/lib.rs [hunk 5]
```diff
@@ -861,6 +866,46 @@ pub mod pallet {
...
+
+			let revokee = T::Lookup::lookup(revokee)?;
+			ClusterManagers::<T>::mutate(|grantees| {
+				if let Some(pos) = grantees.iter().position(|g| g == &revokee) {
+					grantees.remove(pos);
+				}
+			});
+
```
Comment (Raid5594) [inline pallets/ddc-staking/src/lib.rs:906]: let's add an event here and in tests as well

File: pallets/ddc-staking/src/lib.rs [hunk 6]
```diff
@@ -329,6 +329,11 @@ pub mod pallet {
 	#[pallet::getter(fn pricing)]
 	pub type Pricing<T: Config> = StorageValue<_, u128>;
 
+	/// A list of accounts allowed to become cluster managers.
+	#[pallet::storage]
+	#[pallet::getter(fn cluster_managers)]
+	pub type ClusterManagers<T: Config> = StorageValue<_, Vec<T::AccountId>, ValueQuery>;
```
Comment (Raid5594) [inline pallets/ddc-staking/src/lib.rs:335]: General question, how this data structure is integrated into current functionality? what is the goal of a cluster manager?

File: pallets/ddc-staking/src/lib.rs [hunk 7]
```diff
@@ -861,6 +866,46 @@ pub mod pallet {
...
 			Ok(())
 		}
+
+		/// Add a new account to the list of cluster managers.
+		///
+		/// RuntimeOrigin must be Root to call this function.
+		#[pallet::weight(T::WeightInfo::allow_cluster_manager())]
+		pub fn allow_cluster_manager(
+			origin: OriginFor<T>,
+			grantee: <T::Lookup as StaticLookup>::Source,
```
[the remark this answers is not in the document]
Comment (khssnv) [reply pallets/ddc-staking/src/lib.rs:876]: Lookup gives more flexibility to the caller on how they can specify the target.

More details: https://substrate.stackexchange.com/questions/4339/tlookup-as-staticlookupsource-vs-taccountid

File: pallets/ddc-staking/src/lib.rs (hunk 4 above)
Comment (khssnv) [reply pallets/ddc-staking/src/lib.rs:878]: Please check similar thread here: https://github.com/Cerebellum-Network/blockchain-node/pull/82#discussion_r1312721846

File: pallets/ddc-staking/src/lib.rs (hunk 6 above)
Comment (khssnv) [reply pallets/ddc-staking/src/lib.rs:335]: It is a list of accounts that governance allows to become a cluster manager. The list is required for authorization in off-chain systems. Further development will introduce depositing to get into the list.

File: pallets/ddc-staking/src/lib.rs (hunk 3 above)
Comment (khssnv) [reply pallets/ddc-staking/src/lib.rs:886]: Events not required by the user.

File: pallets/ddc-staking/src/lib.rs (hunk 5 above)
Comment (khssnv) [reply pallets/ddc-staking/src/lib.rs:906]: Events not required by the user.

**Review by MRamanenkau — APPROVED** [review]

File: pallets/ddc-staking/src/lib.rs (hunk 5 above)
Comment (Raid5594) [reply pallets/ddc-staking/src/lib.rs:906]: Do you mean by cluster management?

Comment (khssnv) [reply pallets/ddc-staking/src/lib.rs:906]: Yes

Comment (khssnv) [reply pallets/ddc-staking/src/lib.rs:906]: The use does not require tracking live updates or time of the list modifications.

**Pull request merged**